src/main/java/org/openjdk/jextract/json/JSONObjects.java

```diff
@@ -0,0 +1,826 @@
+        return (Map<K, V>) convertToMap(jsonObj, keyType, valueType);
+    }
+
+    /**
+     * Converts the given JSONObject as a EnumMap.
+     *
+     * @param jsonObj the JSONValue being converted.
+     * @param enumCls the target enum key Class for the Map.
+     * @param valueType the value key type for the map.
+     */
+    @SuppressWarnings("unchecked")
+    public static <K extends Enum<K>, V> EnumMap<K, V> toEnumMap(JSONObject jsonObj,
+                Class<K> enumCls, Class<V> valueType) {
+        Objects.requireNonNull(jsonObj);
+        Objects.requireNonNull(enumCls);
+        Objects.requireNonNull(valueType);
+
+        return (EnumMap<K, V>) convertToEnumMap(jsonObj, enumCls, valueType);
+    }
+
+    // package private helpers for JSONArrays
+    @SuppressWarnings("unchecked")
+    static <T> List<T> toList(JSONArray ja, Class<T> elemType) {
+        return (List<T>) convertToList(ja, elemType);
+    }
+
+    @SuppressWarnings("unchecked")
+    static <T> Set<T> toSet(JSONArray ja, Class<T> elemType) {
+        return (Set<T>) convertToSet(ja, elemType);
+    }
+
+    @SuppressWarnings("unchecked")
+    static <T extends Enum<T>> EnumSet<T> toEnumSet(JSONArray ja, Class<T> enumCls) {
+        return (EnumSet<T>) convertToEnumSet(ja, enumCls);
+    }
+
+    // Internals only below this point
+
+    // Type helpers
+
+    // is the given Type Optional raw type or Optional<T>?
+    private static boolean isOptionalType(Type type) {
+        return type == Optional.class ||
+            (type instanceof ParameterizedType pt &&
+            pt.getRawType() == Optional.class);
+    }
+
+    // is this a List/ArrayList/Iterable?
+    private static boolean isListOrIterable(Class<?> cls) {
+        return cls == List.class || cls == ArrayList.class || cls == Iterable.class;
+    }
+
+    // is this a Set/HashSet/LinkedHashSet?
+    private static boolean isSet(Class<?> cls) {
+        return cls == Set.class || cls == HashSet.class || cls == LinkedHashSet.class;
+    }
+
+    // is this a Map/HashMap/LinkedHashMap?
+    private static boolean isMap(Class<?> cls) {
+        return cls == Map.class || cls == HashMap.class || cls == LinkedHashMap.class;
+    }
+
+    // is this String or CharSequence class?
+    private static boolean isStringOrCharSequence(Type type) {
+        return type == String.class || type == CharSequence.class;
+    }
+
+    // The ParamerizedType is expected to have only one type argument at most
+    private static Type getFirstTypeArgument(ParameterizedType pt) {
+        return getNthTypeArgument(pt, 0);
+    }
+
+    private static Type getSecondTypeArgument(ParameterizedType pt) {
+        return getNthTypeArgument(pt, 1);
+    }
+
+    private static Type getNthTypeArgument(ParameterizedType pt, int n) {
+        Type[] typeArgs = pt.getActualTypeArguments();
+        Type nthType = typeArgs.length > n ? typeArgs[n] : Object.class;
+        if (nthType instanceof WildcardType wt) {
```

Comment:
This logic is not correct. E.g. if something is `List<? super Integer>` it is not correct to assume that the type argument is `String` (and e.g. create a `Integer[]` as a result). Because `List<? extends Number>` can have `Number` elements which won't go in an `Integer[]`.

In general, I think it might be appropriate to dial down the level of "smartness" and try to provide something that is well-behaved, and good enough for what we need. We don't need to solve the general problem of mapping json to Java, we just need a mapping that is good enough for jextract to use.

So... I think we'd be ok if wildcards won't be accepted by the mapper.

Reply:
Okay. I'll remove Wildcard types.